src/Scripts/build.ts

```diff
@@ -95,53 +95,7 @@ export async function build(): Promise<string[]> {
             ]
         });
 
-        // Get and store all diagnostic messages generated by typescript to display them when the task
-        // finishes
-        const allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
-        const defaultFormatHost: ts.FormatDiagnosticsHost = {
-            getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
-            getCanonicalFileName: fileName => fileName,
-            getNewLine: () => ts.sys.newLine
-        };
-
-        const formattedDiagnostics = ts.formatDiagnosticsWithColorAndContext(allDiagnostics, defaultFormatHost);
-
-        // If an error causes a compilation failure, display the diagnostic messages and fail
-        if (emitResult.emitSkipped) {
-            process.stdout.write(`\r\x1b[1;31m✖\x1b[0m Failed building ${projectName || 'project'}\n`);
-
-            console.log(formattedDiagnostics);
-
-            throw new Error('Compilation failed.');
-        }
-
-        // Validate thingworx-specific constraints
-        for (const key in twConfig.store) {
-            if (key.startsWith('@')) continue;
-
-            const transformer = twConfig.store[key];
-            transformer.validateConstraints();
-        }
-
-        const diagnosticMessages = twConfig.store['@diagnosticMessages'] as unknown as DiagnosticMessage[];
-
-        // If any errors were reported, display them and fail
-        if (diagnosticMessages.some(m => m.kind == DiagnosticMessageKind.Error)) {
-            process.stdout.write(`\r\x1b[1;31m✖\x1b[0m Failed building ${projectName || 'project'}\n`);
-
-            for (const message of diagnosticMessages) {
-                switch (message.kind) {
-                    case DiagnosticMessageKind.Error:
-                        console.log(`🛑 \x1b[1;31mError\x1b[0m ${message.message}`);
-                        break;
-                    case DiagnosticMessageKind.Warning:
-                        console.log(`🔶 \x1b[1;33mWarning\x1b[0m ${message.message}`);
-                        break;
-                }
-            }
-
-            throw new Error('Validation failed.');
-        }
+        const { tsDiagnostics, validationDiagnostics } = firePostTransformActions(program, twConfig.store, projectName, emitResult);
```

Comment:
Not really a fan of this, because this function is pretty tightly coupled with the build process (e.g. it outputs messages like ```process.stdout.write(`\r\x1b[1;31m✖\x1b[0m Failed building ${projectName || 'project'}\n`);``` which require the previous message to be printed before invoking it), but also because it displays some diagnostic messages in some cases, but in other cases it's still the build function that has to display them.

I would just extract out the parts that are strictly needed for the export api task and use them there instead.

Reply:
Well, the `firePostTransformActions` should be fired on every transformer in`generate-api` because we need to handle the datashape inheritance.
Since we fire it, and it generates warnings and such, shouldn't this warnings be displayed to the user?